x-pack/plugins/security_solution/public/detection_engine/rule_creation/logic/esql_validator.ts

```diff
@@ -34,11 +38,39 @@ const constructValidationError = (error: Error) => {
   };
 };
 
+const constructSyntaxError = (error: Error) => {
+  return {
+    code: ERROR_CODES.INVALID_SYNTAX,
+    message: error?.message
+      ? i18n.esqlValidationErrorMessage(error.message)
+      : i18n.ESQL_VALIDATION_UNKNOWN_ERROR,
+    error,
+  };
+};
+
 /**
  * checks whether query has metadata _id operator
  */
-export const computeHasMetadataOperator = (esqlQuery: string) => {
-  return /(?<!\|[\s\S.]*)\s*metadata[\s\S.]*_id[\s\S.]*/i.test(esqlQuery?.split('|')?.[0]);
+export const computeHasMetadataOperator = (ast: ESQLAst) => {
+  const fromCommand = ast.find((astItem) => astItem.type === 'command' && astItem.name === 'from');
+
+  // Check whether the `from` command has `metadata` operator
+  const metadataOption = fromCommand?.args.find(
+    (fromArg) => isOptionItem(fromArg) && fromArg.name === 'metadata'
+  );
+  if (!metadataOption) {
+    return false;
+  }
+
+  // Check whether the `metadata` operator has `_id` argument
+  const idColumnItem = (metadataOption as ESQLCommandOption).args.find(
```

Comment:
why casting is needed here? 
Given that one of the types there does not have `args` property, it could lead to error of accessing property. Something we had already in the past.

Reply:
Seems like Array.find cannot infer the type in this case. I re-arranged the code.